### DISCUSSION TITLE
How to select recursively for parent-children relationship?

### DISCUSSION THREAD

**Discussion by richardkyk** [supabase/supabase#2588] [Questions] [answered]
I'm fairly new to Supabase and PostgreSQL but so far I'm really enjoying the experience!

I'm trying to get a nested tree structure for a single group to visualise an organisation chart.

I have a table called "groups" with a foreign key (column called "parent") that points back to the same table. 

The following code works, however it only gets the immediate children and doesn't get the grandchildren etc.

How can I get the nested trees for all the root nodes to select recursively? 

```javascript
const { data, error } = await supabase
      .from("groups")
      .select(
        `
        id, name, parent, children:groups(*)
        `
      )
      .is("parent", null);
```

**Comment by steve-chavez** [comment, marked as answer]
Hey @richardkyk,

Purely on the js client, you'd have to explicitly specify how many levels you want:

```js
const { data, error } = await supabase
      .from("groups")
      .select(
        `
        id, name, parent, children:groups(*,
          children:groups(*,
            children:groups(*)
          )
        )
        `
      )
      .is("parent", null);
```

However, there's a way to do this recursively at once - by using a SQL [recursive](https://www.postgresqltutorial.com/postgresql-recursive-query/) view. Then calling this view with a `supabase.from('my_view')`.

Here you can see an example of how this is done for [threaded comments](https://github.com/lawrencecchen/threaded-comments/blob/main/sql/comments_view.sql#L87-L167).

**Reply by richardkyk** [reply]
Hi @steve-chavez,

Thanks for the reply! I attempted a solution using the examples:
```
create recursive view groups_tree (
  id, name, parent, members
  ) as
  select
    id, name, parent, members
  from
    groups
  where
    parent is null
  union all
  select
    g.id, g.name, g.parent, g.members
  from
    groups g
  inner join groups_tree gt on gt.id = g.parent;



create or replace view groups_with_children as
  select distinct on (id)
    id,
    name,
    parent,
    members,
    children
from groups_tree;
```

And then calling it with
```javascript
const { data, error } = await supabase
      .from("groups_with_children")
```

It returns the data as a flat array, is there a way to get SQL to return the nested structure similar to what the js client returns? Or will I need to construct the tree on the front end.
